crates/sui-framework/packages/move-stdlib/sources/vector.move

```diff
@@ -361,4 +395,49 @@ module std::vector {
         zip_do_ref!($v1, $v2, |el1, el2| r.push_back($f(el1, el2)));
         r
     }
+
+    /// Iterate through the vector and find the indices of elements that satisfy the predicate `p`.
+    /// Returns a new vector containing the indices of elements that satisfy the predicate in the
+    /// order they appear in the original vector.
+    public macro fun find_indices<$T>($v: &vector<$T>, $p: |&$T| -> bool): vector<u64> {
+        let v = $v;
+        let mut r = vector[];
+        v.length().do!(|i| if ($p(&v[i])) r.push_back(i));
+        r
+    }
+
+    /// Takes the first `n` elements of the vector `v` that satisfy the predicate `p`.
+    /// Modifies the original vector.
+    public macro fun take_mut_while<$T>($v: &mut vector<$T>, $p: |&$T| -> bool): vector<$T> {
+        let v = $v;
+        let mut r = vector[];
+        v.reverse();
+        while (v.length() != 0 && $p(&v[v.length() - 1])) {
+            r.push_back(v.pop_back());
+        };
+        v.reverse();
+        r
+    }
+
+    /// Takes the first `n` elements of the vector `v` that satisfy the predicate `p` and drops the rest.
+    /// Destroys the original vector after taking the elements.
+    public macro fun take_while<$T: drop>($v: vector<$T>, $p: |&$T| -> bool): vector<$T> {
+        let mut v = $v;
+        take_mut_while!(&mut v, $p)
+    }
+
+    /// Take all elements of the vector `v` except the first `n` elements that satisfy the predicate `p`.
+    /// Modifies the original vector.
+    public macro fun skip_mut_while<$T>($v: &mut vector<$T>, $p: |&$T| -> bool): vector<$T> {
+        let v = $v;
+        // find first element that does not satisfy the predicate, then skip from there
+        v.find_index!(|e| !$p(e)).map!(|c| v.skip_mut(c)).destroy_or!(vector[])
```

Comment:
Nit, while slick, it might be nice to avoid the 2 separate iterations over `v`. Especially since this is library code, small optimizations can be really meaningful